transport-native-epoll/src/main/c/netty_epoll_linuxsocket.c

```diff
@@ -262,6 +267,134 @@ static jobject netty_epoll_linuxsocket_getPeerCredentials(JNIEnv *env, jclass cl
+
+    // If a packet was received...
+    if (res >= (jlong) sizeof(packetInfo)) {
+
+        // If the packet was truncated because the receive buffer was too small,
+        // raise an IOException containing the error ENOMEM.
+        if ((packetInfo.flags & TUN_PKT_STRIP) != 0) {
+            netty_unix_errors_throwIOExceptionErrorNo(env, "readTunTapPacketAddress() failed: ", ENOMEM);
+            res = -1;
+        }
+
+        // Otherwise return the protocol type in the upper 16 bits and the length
+        // of the packet (not including the packet info header) in the lower 32 bits.
+        else {
+            jlong proto = htons(packetInfo.proto);
+            jlong packetLen = res - sizeof(packetInfo);
+            res = (proto << 32) | packetLen;
+        }
+    }
+
+    // If an error occurred, return to the caller empty-handed if the kernel signaled
+    // that no packet was available.  Otherwise, raise an IOException containing the error code.
+    else if (res < 0) {
+        if (errno == EAGAIN || errno == EWOULDBLOCK) {
+            res = 0;
+        }
+        else {
+            netty_unix_errors_throwIOExceptionErrorNo(env, "readTunTapPacketAddress() failed: ", errno);
+        }
+    }
+
+    // If for some reason the kernel returned less bytes than are needed to
+    // convey the packet info header, treat it as if no packet had been received.
+    // (Note that it's not clear the kernel will ever do this in practice,
+    // therefore this code is purely defensive).
+    else {
+        res = 0;
+    }
+
+    return res;
+}
+
+static jlong netty_epoll_linuxsocket_readTunTapPacket(JNIEnv *env, jclass cls, jint fd, jobject jbuf, jint pos, jint limit) {
+    void *buf = (*env)->GetDirectBufferAddress(env, jbuf);
+    return netty_epoll_linuxsocket_readTunTapPacketAddress(env, cls, fd, (jlong)buf, pos, limit);
+}
+
+static void netty_epoll_linuxsocket_writeTunTapPacketAddress(JNIEnv *env, jclass cls, jint fd, jint protocol, jlong addr, jint pos, jint limit) {
+    jint res;
+    struct iovec iov[2];
+    struct tun_pi packetInfo;
+
+    memset(&packetInfo, 0, sizeof(packetInfo));
+    packetInfo.flags = 0;
+    packetInfo.proto = htons(protocol & 0xFFFF);
+
+    iov[0].iov_base = (void *)&packetInfo;
+    iov[0].iov_len = sizeof(packetInfo);
+    iov[1].iov_base = ((char *)addr) + pos;
+    iov[1].iov_len = limit;
+
+    // Attempt to write the packet. Loop if the write is interrupted by a signal.
+    do {
+        res = (jint)writev((int)fd, iov, 2);
+    } while (res < 0 && errno == EINTR);
+
+    // If an error occurred raise an IOException containing the error code.
+    if (res < 0) {
+        netty_unix_errors_throwIOExceptionErrorNo(env, "writeTunTapPacketAddress() failed: ", errno);
+    }
+}
+
+static void netty_epoll_linuxsocket_writeTunTapPacket(JNIEnv *env, jclass cls, jint fd, jint protocol, jobject jbuf, jint pos, jint limit) {
+    void *buf = (*env)->GetDirectBufferAddress(env, jbuf);
```

Comment:
It looks like this may be called when the buffer doesn't have an address ... I made a suggestion/comment in he java code on this topic about how we typically handle this case.